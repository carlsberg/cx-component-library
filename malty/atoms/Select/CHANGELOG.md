# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- added missing theme context ([#366](https://github.com/CarlsbergGBS/cx-component-library/issues/366)) ([03955fb](https://github.com/CarlsbergGBS/cx-component-library/commit/03955fb08d99aef4437323eebd0b018bb044916a))
- **checkbox:** [DSM-397] Refactor checkbox component ([#551](https://github.com/CarlsbergGBS/cx-component-library/issues/551)) ([82eb0a6](https://github.com/CarlsbergGBS/cx-component-library/commit/82eb0a62c27c1df768ae94643365942d5f4398e1))
- **DSM-257:** full width & update state when options change ([#244](https://github.com/CarlsbergGBS/cx-component-library/issues/244)) ([7b63525](https://github.com/CarlsbergGBS/cx-component-library/commit/7b635253f57759d297a26cf15e23b159a2724b46))
- **DSM-331:** fixed theme issue ([#290](https://github.com/CarlsbergGBS/cx-component-library/issues/290)) ([325dd37](https://github.com/CarlsbergGBS/cx-component-library/commit/325dd371de145534d46e0fa47a1d5c63a169e6cb))
- Remove `eslint` warnings and errors ([#322](https://github.com/CarlsbergGBS/cx-component-library/issues/322)) ([4c7bc73](https://github.com/CarlsbergGBS/cx-component-library/commit/4c7bc737de8abb060919ac5ba16220a6d5d0ace9))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Select min-width & translatable props ([#344](https://github.com/CarlsbergGBS/cx-component-library/issues/344)) ([72354d9](https://github.com/CarlsbergGBS/cx-component-library/commit/72354d95a0fbf12932340a1fe48b19943c26b0bd))
- **select-component:** Select button default type ([#154](https://github.com/CarlsbergGBS/cx-component-library/issues/154)) ([aff5844](https://github.com/CarlsbergGBS/cx-component-library/commit/aff5844a99d7e9c99530ba3de2bbec829fcdc5a5))
- **Select:** [DSM-564] Add required prop ([#578](https://github.com/CarlsbergGBS/cx-component-library/issues/578)) ([ae44c4c](https://github.com/CarlsbergGBS/cx-component-library/commit/ae44c4cce533f911503620f050a3cd0aa594f79f))
- **select:** [DSM-764] update options description ([#759](https://github.com/CarlsbergGBS/cx-component-library/issues/759)) ([5cca3c9](https://github.com/CarlsbergGBS/cx-component-library/commit/5cca3c92542f46f3a3d68ca61cc893f74dc33c7d))
- **select:** Update value from outside when set an empty array ([#321](https://github.com/CarlsbergGBS/cx-component-library/issues/321)) ([8033a7a](https://github.com/CarlsbergGBS/cx-component-library/commit/8033a7ad4917792640fa0766f0f599da4da05856))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- AlertInline and Toast ([#182](https://github.com/CarlsbergGBS/cx-component-library/issues/182)) ([8f344e3](https://github.com/CarlsbergGBS/cx-component-library/commit/8f344e392a6b339d3f159c3e8a2958ef5ddabd71))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-204:** dropdown search ([#155](https://github.com/CarlsbergGBS/cx-component-library/issues/155)) ([60263ac](https://github.com/CarlsbergGBS/cx-component-library/commit/60263ac4e403a0040f86c1f7059ae27a5b88f14b))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-235:** added readOnly on inputs ([#181](https://github.com/CarlsbergGBS/cx-component-library/issues/181)) ([36c5ef7](https://github.com/CarlsbergGBS/cx-component-library/commit/36c5ef7cb1225bf5a150918e01e10fd253377110))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-285:** table ([#442](https://github.com/CarlsbergGBS/cx-component-library/issues/442)) ([8f81f8d](https://github.com/CarlsbergGBS/cx-component-library/commit/8f81f8d3436af86fd023d9737245a76a142ad51f))
- **DSM-352:** clear button on single select ([#301](https://github.com/CarlsbergGBS/cx-component-library/issues/301)) ([479b7ea](https://github.com/CarlsbergGBS/cx-component-library/commit/479b7ea10b8cf11d521414aa1d7d8b9b6e3e7aa2))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-405:** expose onBlur event ([#346](https://github.com/CarlsbergGBS/cx-component-library/issues/346)) ([43a84d6](https://github.com/CarlsbergGBS/cx-component-library/commit/43a84d6c9dacbe6ee65b94e20302017e9b45cc1b))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **DSM-515:** product card ([#548](https://github.com/CarlsbergGBS/cx-component-library/issues/548)) ([9b91827](https://github.com/CarlsbergGBS/cx-component-library/commit/9b918271c05d4a98fdd7c5b34bb356f89a92c991))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- **MS-111:** prop to align dropdown to right and remove clear all option ([#404](https://github.com/CarlsbergGBS/cx-component-library/issues/404)) ([358d1e2](https://github.com/CarlsbergGBS/cx-component-library/commit/358d1e2cca4793a47f429881fe1589f8ea2eb1af))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- **Select:** [DSM-494] Disable case sensitive search on Select ([#591](https://github.com/CarlsbergGBS/cx-component-library/issues/591)) ([98cd84c](https://github.com/CarlsbergGBS/cx-component-library/commit/98cd84c124762385aff4c24c25f42707cc23f5cd))
- **select:** [DSM-731] allow opening over other components ([#760](https://github.com/CarlsbergGBS/cx-component-library/issues/760)) ([705d842](https://github.com/CarlsbergGBS/cx-component-library/commit/705d842c41a0b0f5501a7060cdfe972a4810d09b))
- **select:** [DSM-758] add tests ([#743](https://github.com/CarlsbergGBS/cx-component-library/issues/743)) ([df5a37a](https://github.com/CarlsbergGBS/cx-component-library/commit/df5a37aa9cd5d408d798f3a977026737a9fae240))
- **select:** [MS-197] add ellipsis on select options ([#857](https://github.com/CarlsbergGBS/cx-component-library/issues/857)) ([2b6cec6](https://github.com/CarlsbergGBS/cx-component-library/commit/2b6cec65cdaa0dd96212f4821f53dea035b36563))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
