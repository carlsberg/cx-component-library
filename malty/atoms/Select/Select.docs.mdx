---
labels: ['react', 'typescript', 'select']
description: 'A Select component.'
---
