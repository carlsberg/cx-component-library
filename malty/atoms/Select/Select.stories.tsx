import { generateStorybookSpacing } from '@carlsberggroup/malty.utils.space';
import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import { Select as SelectComponent, SelectOptionsType, SelectProps, SelectSize, SelectType } from '.';

const StyledContainer = styled.div`
  display: flex;
  justify-content: center;
  height: 300px;
`;

const meta: Meta<SelectProps> = {
  title: 'Forms/Select',
  component: SelectComponent,
  render: (args) => {
    return (
      <StyledContainer>
        <SelectComponent {...args} />
      </StyledContainer>
    );
  },
  parameters: {
    importObject: 'Select',
    importPath: '@carlsberggroup/malty.atoms.select'
  },
  argTypes: {
    label: {
      description: 'Label for the Select component',
      control: 'text'
    },
    selectAllLabel: {
      description: 'Label for the Select all action',
      control: 'text'
    },
    clearAllLabel: {
      description: 'Label for the Clear all action',
      control: 'text'
    },
    placeholder: {
      description: 'Placeholder text to go inside the Select field, when empty.',
      control: 'text'
    },
    error: {
      description: 'Error message to be displayed when error is present.',
      control: 'text'
    },
    hint: {
      description: 'helper message to be displayed',
      control: 'text'
    },
    size: {
      description: 'Select component size.',
      options: Object.values(SelectSize),
      control: {
        type: 'radio'
      }
    },
    type: {
      description: 'Type of select component',
      options: Object.values(SelectType),
      control: {
        type: 'select'
      }
    },
    disabled: {
      description: 'Select state, disabled.',
      control: 'boolean'
    },
    readOnly: {
      control: 'boolean',
      description: 'Select state, readOnly.'
    },
    defaultValue: {
      description: 'Initial selected option'
    },
    options: {
      description:
        'This is the collection of possible options on the select, keep in mind that value will be evaluated as case-sensitive'
    },
    multiple: {
      description: 'Select state, when active allows for multi selection',
      control: 'boolean'
    },
    search: {
      description: 'Select state, when active allows for multi selection',
      control: 'boolean'
    },
    selectionText: {
      description: 'Text to display when multiple options are selected. Ex. 3 "options selected" ',
      control: 'text'
    },
    onValueChange: {
      description: 'Function to be executed when an option is selected or unselected'
    },
    dataTestId: {
      control: 'text',
      description: 'select data-testid'
    },
    required: {
      description: 'Makes the select visually mandatory to be filled when not in inline version',
      control: 'boolean'
    },
    optionsZIndex: {
      control: 'number',
      description: 'Controls the z-index of the options wrapper'
    },
    ...generateStorybookSpacing()
  }
};

const testOptions: SelectOptionsType[] = [
  {
    value: 'value 1',
    name: 'name 1'
  },
  {
    value: 'value 2',
    name: 'name 2'
  },
  {
    value: 'value 3',
    name: 'name 3'
  },
  {
    value: 'value 4',
    name: 'name 4'
  },
  {
    value: 'value 5',
    name: 'name 5'
  }
];

type Story = StoryObj<SelectProps>;

export const Base: Story = {
  args: {
    options: testOptions,
    size: SelectSize.Medium,
    label: 'Label',
    type: SelectType.Default,
    hint: 'hint text',
    disabled: false,
    placeholder: 'Placeholder',
    defaultValue: [testOptions[0]],
    selectionText: 'options selected',
    readOnly: false,
    required: false,
    optionsZIndex: 2,
    onValueChange: action('onValueChange')
  }
};

export const Inline: Story = {
  args: {
    options: testOptions,
    size: SelectSize.Medium,
    label: 'Label',
    type: SelectType.Inline,
    hint: 'hint text',
    disabled: false,
    placeholder: 'Placeholder',
    defaultValue: [testOptions[0]],
    selectionText: 'options selected',
    readOnly: false,
    required: false,
    optionsZIndex: 2,
    onValueChange: action('onValueChange')
  }
};

export default meta;
