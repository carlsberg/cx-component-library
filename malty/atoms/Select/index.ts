export { Select } from './Select';
export { SelectSize, SelectType } from './Select.types';
export type { SelectOptionsType, SelectProps } from './Select.types';
