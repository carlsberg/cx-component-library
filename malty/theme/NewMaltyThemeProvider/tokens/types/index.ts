export type { PrimitiveBorderRadiusTokens, SemanticBorderRadiusTokens } from './borderRadius.types';
export type { PrimitiveBorderTokens, SemanticBorderTokens } from './borders.types';
export type { PrimitiveColorsTokens, SemanticColorsTokens } from './colors.types';
export type { GridTokens } from './grid.types';
export type { OpacityTokens } from './opacity.types';
export type { ShadowTokens } from './shadows.types';
export type {
  AdditionalPrimitives,
  BorderRadiusTokens,
  ColorTokens,
  SemanticOverrides,
  SemanticTokens
} from './shared.types';
export type { SizeTokens } from './sizes.types';
export type { TypographyTokens } from './typography.types';
