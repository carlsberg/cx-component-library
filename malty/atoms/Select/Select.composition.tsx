import React, { useState } from 'react';
import { Select } from './Select';
import { SelectType } from './Select.types';

export const SelectText = () => {
  const [value, setValue] = useState('');
  return (
    <Select
      options={}
      type={SelectType.Dropdown}
      value={value}
      onValueChange={(newValue: string) => setValue(newValue)}
    />
  );
};
