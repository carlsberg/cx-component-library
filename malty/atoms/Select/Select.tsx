import { Checkbox } from '@carlsberggroup/malty.atoms.checkbox';
import { IconColors, IconSizesTypes } from '@carlsberggroup/malty.atoms.icon';
import { globalTheme as defaultTheme, TypographyProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useEffect, useMemo, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { v4 as uuid } from 'uuid';
import {
  StyledButton,
  StyledButtonContainer,
  StyledCheck,
  StyledChevronDown,
  StyledError,
  StyledLabel,
  StyledOption,
  StyledOptionsWrapper,
  StyledSuccess,
  StyledWrapper
} from './Select.styled';
import { OptionsType, SelectProps, SelectType, SizeTypes } from './Select.types';

export const Select = ({
  defaultValue,
  onValueChange,
  options,
  placeholder,
  label,
  type,
  error,
  success,
  disabled = false,
  size = SizeTypes.Medium,
  children,
  multiple = false,
  className
}: SelectProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;
  const id = useMemo(() => uuid(), []);
  // eslint-disable-next-line radix
  const [numSize, setNumSize] = useState(parseInt(theme.variables.select.size.medium.value));
  const [showOptionList, setShowOptionList] = useState(false);
  const [selectedValueState, setSelectedValueState] = useState(defaultValue || []);

  const toggleOptionList = () => {
    setShowOptionList(!showOptionList);
  };

  const handleOptionSelected = (option: OptionsType) => {
    // eslint-disable-next-line no-restricted-globals
    onValueChange(option);
    // eslint-disable-next-line no-restricted-globals
    event?.preventDefault();
    if (!multiple) {
      setShowOptionList(false);
    }
    if (checkIfSelected(option)) {
      if (multiple) {
        removeSelectedOption(option);
      }
    } else {
      if (!multiple) {
        selectedValueState.pop();
      }
      // setSelectedValueState([...selectedValueState, option]);
      setSelectedValueState((prev) => [...prev, option]);
    }
  };

  const checkIfSelected = (option: OptionsType) =>
    !!selectedValueState.find((el: OptionsType) => el.value === option.value);

  const removeSelectedOption = (option: OptionsType) => {
    const index = selectedValueState?.indexOf(option);
    const auxSelected = JSON.parse(JSON.stringify(selectedValueState));
    auxSelected?.splice(index, 1);
    update(auxSelected);
  };

  const update = (auxSelected: OptionsType[]) => {
    setSelectedValueState(auxSelected);
    // setSelectedValueState(option);
  };

  useEffect(() => {
    switch (size) {
      case SizeTypes.Large: {
        // eslint-disable-next-line radix
        setNumSize(parseInt(theme.variables.select.size.large.value));
        break;
      }
      default: {
        // eslint-disable-next-line radix
        setNumSize(parseInt(theme.variables.select.size.medium.value));
        break;
      }
    }
  }, [size, theme]);

  const renderDefaultDropdown = () => (
    <TypographyProvider>
      {showOptionList && (
        <StyledOptionsWrapper selectStyle={type} theme={theme} height={numSize}>
          {children}
          {!children &&
            options?.map((option: OptionsType, index: number) => (
              <StyledOption
                theme={theme}
                key={option.value}
                value={option.value}
                onClick={() => handleOptionSelected(option)}
                height={numSize}
                selected={!!selectedValueState.find((el: OptionsType) => el.value === option.value)}
                selectStyle={type}
                disabled={disabled}
                data-testid={`select-option-${index}`}
                className={className}
              >
                {multiple && (
                  <Checkbox
                    labelText={option.name as string}
                    value={option.value}
                    onValueChange={() => null}
                    checked={!!selectedValueState.find((el: OptionsType) => el.value === option.value)}
                  />
                )}
                {!multiple && (
                  <>
                    <StyledWrapper>
                      {option.icon}
                      {option.name}
                    </StyledWrapper>
                    {selectedValueState.find((el: OptionsType) => el.value === option.value) && (
                      <StyledCheck
                        selectStyle={type}
                        color={IconColors.Primary}
                        size={size === SizeTypes.Large ? IconSizesTypes.Medium : IconSizesTypes.Small}
                      />
                    )}
                  </>
                )}
              </StyledOption>
            ))}
        </StyledOptionsWrapper>
      )}
    </TypographyProvider>
  );
  return (
    <TypographyProvider>
      {label && type !== SelectType.Inline && (
        <StyledLabel disabled={disabled} htmlFor={id} theme={theme}>
          {label}
        </StyledLabel>
      )}
      <StyledButtonContainer selectStyle={type} theme={theme}>
        <StyledButton
          name={id}
          id={id}
          theme={theme}
          isActive={selectedValueState?.length > 0 || type === SelectType.Inline}
          selectStyle={type}
          height={numSize}
          onClick={() => toggleOptionList()}
          disabled={disabled}
          isError={!!error && type !== SelectType.Inline}
          isSuccess={!!success && type !== SelectType.Inline}
          open={showOptionList}
        >
          {selectedValueState.length > 0
            ? selectedValueState?.map((selectedValue: OptionsType, index: number) =>
                index !== 0 ? `, ${selectedValue.name}` : selectedValue.name
              )
            : (type === SelectType.Inline && label) || (type === SelectType.Default && placeholder)}

          <StyledChevronDown
            selectStyle={type}
            disabled={disabled}
            open={showOptionList}
            color={IconColors.Primary}
            size={IconSizesTypes.Medium}
          />
        </StyledButton>

        {renderDefaultDropdown()}
      </StyledButtonContainer>

      {error && type !== SelectType.Inline && <StyledError theme={theme}>{error}</StyledError>}
      {success && !error && type !== SelectType.Inline && <StyledSuccess theme={theme}>{success}</StyledSuccess>}
    </TypographyProvider>
  );
};
