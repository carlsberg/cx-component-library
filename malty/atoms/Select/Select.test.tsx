import { jsonRenderer, render } from '@carlsberggroup/malty.utils.test';
import { screen } from '@testing-library/react';
import React from 'react';
import { Select } from './Select';
import { SelectSize, SelectType } from './Select.types';

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

const mockFn = jest.fn();
const testOptions = [
  {
    value: 'value 1',
    name: 'name 1'
  },
  {
    value: 'value 2',
    name: 'name 2'
  },
  {
    value: 'value 3',
    name: 'name 3'
  },
  {
    value: 'value 4',
    name: 'name 4'
  },
  {
    value: 'value 5',
    name: 'name 5'
  }
];
describe('select', () => {
  it('matches snapshot', () => {
    const view = jsonRenderer(
      <Select
        options={testOptions}
        label="Label text"
        placeholder="Placeholder text"
        onValueChange={mockFn}
        type={SelectType.Default}
        size={SelectSize.Medium}
      />
    );
    expect(view).toMatchSnapshot();
  });

  it('renders elements', () => {
    render(
      <Select
        options={testOptions}
        label="Label text"
        placeholder="Placeholder text"
        onValueChange={mockFn}
        type={SelectType.Default}
        error="Error text"
        size={SelectSize.Medium}
      />
    );
    expect(screen.getByLabelText('Label text')).toBeInTheDocument();
    expect(screen.getByText('Placeholder text')).toBeInTheDocument();
    expect(screen.getByText('Error text')).toBeInTheDocument();
  });

  it('calls onValueChange on click', () => {
    const onValueChange = jest.fn();
    const { rerender } = render(
      <Select
        options={testOptions}
        label="select label"
        placeholder="select"
        onValueChange={onValueChange}
        type={SelectType.Default}
        size={SelectSize.Medium}
        dataTestId="select"
      />
    );
    const select = screen.getByText('select');
    expect(select).toBeVisible();
    // TODO: Replace with click event
    select.click();
    const selectOption = screen.getByTestId('select-option-1');
    expect(selectOption).toBeVisible();
    // TODO: Replace with click event
    selectOption.click();
    expect(onValueChange).toHaveBeenCalledTimes(1);
    rerender(
      <Select
        options={testOptions}
        defaultValue={[testOptions[1]]}
        label="Input label"
        onValueChange={onValueChange}
        type={SelectType.Default}
        size={SelectSize.Medium}
        dataTestId="select"
      />
    );
    const selectedValue = screen.getByTestId('select-selected-values');
    expect(selectedValue.innerHTML).toEqual(testOptions[1].name);
  });

  it('renders inline select', () => {
    const onValueChange = jest.fn();
    render(<Select options={testOptions} label="inline" onValueChange={onValueChange} type={SelectType.Inline} />);
    expect(screen.getByText('inline')).toBeInTheDocument();
  });
});
