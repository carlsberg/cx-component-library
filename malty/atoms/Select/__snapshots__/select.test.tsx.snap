// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select matches snapshot 1`] = `
<span
  className="sc-bczRLJ sc-bZnhIo hROJNc iRtOTt"
>
  <label
    className="sc-ivTmOn fYXiZm"
    data-testid="undefined-label"
    disabled={false}
    htmlFor="00000000-0000-0000-0000-000000000000"
  >
    Label text
  </label>
  <div
    className="sc-bBrHrO eFkHu"
  >
    <button
      className="sc-iIPllB drgmfS"
      data-testid="undefined-button"
      disabled={false}
      height="40px"
      id="00000000-0000-0000-0000-000000000000"
      name="00000000-0000-0000-0000-000000000000"
      onClick={[Function]}
      open={false}
      type="button"
    >
      <div
        className="sc-himrzO cWxDEl"
        data-testid="undefined-selected-values"
      >
        Placeholder text
      </div>
      <svg
        className="sc-gsnTZi gIwTyE sc-kLLXSd eAPGvz"
        color="#212833"
        data-testid="svg-component"
        size="24px"
        version="1.1"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <g
          fillRule="evenodd"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 13.586L6.707 8.293a1 1 0 00-1.414 1.414l6 6a1 1 0 001.414 0l6-6a1 1 0 10-1.414-1.414z"
          />
        </g>
      </svg>
    </button>
    <span
      className="sc-bczRLJ hROJNc"
    >
      <ul
        className="sc-gicCDI jOnkhs"
        data-testid="undefined-options"
        height="40px"
      >
        <div
          className="sc-gXmSlM"
        />
        <li
          className="sc-bZkfAO cqTeUr"
          data-testid="undefined-option-0"
          disabled={false}
          height="40px"
          onClick={[Function]}
          selected={false}
          value="value 1"
        >
          <div
            className="sc-jIZahH jKzfhC"
            data-testid="undefined-option-label-0"
          >
            name 1
          </div>
        </li>
        <li
          className="sc-bZkfAO cqTeUr"
          data-testid="undefined-option-1"
          disabled={false}
          height="40px"
          onClick={[Function]}
          selected={false}
          value="value 2"
        >
          <div
            className="sc-jIZahH jKzfhC"
            data-testid="undefined-option-label-1"
          >
            name 2
          </div>
        </li>
        <li
          className="sc-bZkfAO cqTeUr"
          data-testid="undefined-option-2"
          disabled={false}
          height="40px"
          onClick={[Function]}
          selected={false}
          value="value 3"
        >
          <div
            className="sc-jIZahH jKzfhC"
            data-testid="undefined-option-label-2"
          >
            name 3
          </div>
        </li>
        <li
          className="sc-bZkfAO cqTeUr"
          data-testid="undefined-option-3"
          disabled={false}
          height="40px"
          onClick={[Function]}
          selected={false}
          value="value 4"
        >
          <div
            className="sc-jIZahH jKzfhC"
            data-testid="undefined-option-label-3"
          >
            name 4
          </div>
        </li>
        <li
          className="sc-bZkfAO cqTeUr"
          data-testid="undefined-option-4"
          disabled={false}
          height="40px"
          onClick={[Function]}
          selected={false}
          value="value 5"
        >
          <div
            className="sc-jIZahH jKzfhC"
            data-testid="undefined-option-label-4"
          >
            name 5
          </div>
        </li>
      </ul>
    </span>
  </div>
</span>
`;
